io.atomix.raft.SingleRaftEntryValidationTest#shouldNotAppendInvalidEntryToLog is flaky

**Summary**

- How often does the test fail? So far 3 times in 2 days
- Does it block your work? no, but it makes it longer to merge stuff
- Do we suspect that it is a real failure? no idea yet

**Failures**

<details><summary>Example assertion failure</summary>
<pre>
java.util.concurrent.ExecutionException: io.atomix.raft.RaftException$NoLeader: LeaderRole is closed and cannot be used as appender
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:395)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2022)
	at io.atomix.raft.RaftRule$TestAppendListener.awaitCommit(RaftRule.java:623)
	at io.atomix.raft.RaftRule.appendEntry(RaftRule.java:523)
	at io.atomix.raft.RaftRule.appendEntry(RaftRule.java:516)
	at io.atomix.raft.SingleRaftEntryValidationTest.shouldNotAppendInvalidEntryToLog(SingleRaftEntryValidationTest.java:55)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
	at org.junit.vintage.engine.execution.RunnerExecutor.execute(RunnerExecutor.java:43)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
	at org.junit.vintage.engine.VintageTestEngine.executeAllChildren(VintageTestEngine.java:82)
	at org.junit.vintage.engine.VintageTestEngine.execute(VintageTestEngine.java:73)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:108)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:96)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:75)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.lambda$execute$1(JUnitPlatformProvider.java:199)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.execute(JUnitPlatformProvider.java:193)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invokeAllTests(JUnitPlatformProvider.java:154)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invoke(JUnitPlatformProvider.java:120)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:428)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:562)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:548)
Caused by: io.atomix.raft.RaftException$NoLeader: LeaderRole is closed and cannot be used as appender
	at io.atomix.raft.roles.LeaderRole.safeAppendEntry(LeaderRole.java:565)
	at io.atomix.raft.roles.LeaderRole.lambda$appendEntry$8(LeaderRole.java:552)
	at io.atomix.utils.concurrent.SingleThreadContext$WrappedRunnable.run(SingleThreadContext.java:188)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
</pre>
</details>

**Hypotheses**

None yet, haven't looked at it.

**Logs**

<details><summary>Logs</summary>
<pre>
Standard Output
20:01:13.369 [] TRACE io.atomix.utils.serializer.CompatibleKryoPool - Skipping class java.util.ArrayList: 36 already registered as 517.
20:01:13.370 [] TRACE io.atomix.utils.serializer.CompatibleKryoPool - Skipping class java.util.LinkedList: 37 already registered as 518.
20:01:13.370 [] TRACE io.atomix.utils.serializer.CompatibleKryoPool - Skipping class java.util.Collections$EmptyList: 50 already registered as 519.
20:01:13.371 [] TRACE io.atomix.utils.serializer.CompatibleKryoPool - Skipping class java.util.HashSet: 38 already registered as 520.
20:01:13.387 [] TRACE io.atomix.utils.serializer.CompatibleKryoPool - Skipping class java.util.ArrayList: 36 already registered as 602.
20:01:13.387 [] TRACE io.atomix.utils.serializer.CompatibleKryoPool - Skipping class java.util.HashSet: 38 already registered as 603.
20:01:13.588 [] DEBUG io.zeebe.journal.file.SegmentedJournal - Created segment: JournalSegment{id=1, version=1, index=1}
20:01:13.694 [] TRACE io.atomix.raft.storage.system.MetaStore - Store configuration Configuration{index=0, time=2021-03-28 08:01:13,618, members=[DefaultRaftMember{id=1, type=ACTIVE, updated=2021-03-28T20:01:13.618113Z}]}
20:01:13.709 [] INFO  io.atomix.raft.impl.RaftContext - RaftServer{1} - Transitioning to FOLLOWER
20:01:13.717 [] INFO  io.atomix.raft.roles.FollowerRole - RaftServer{1}{role=FOLLOWER} - Single member cluster. Transitioning directly to candidate.
20:01:13.718 [] INFO  io.atomix.raft.impl.RaftContext - RaftServer{1} - Transitioning to CANDIDATE
20:01:13.777 [] INFO  io.atomix.raft.roles.CandidateRole - RaftServer{1}{role=CANDIDATE} - Single member cluster. Transitioning directly to leader.
20:01:13.777 [] TRACE io.atomix.raft.storage.system.MetaStore - Store term 1
20:01:13.782 [] TRACE io.atomix.raft.storage.system.MetaStore - Store vote null
20:01:13.785 [] DEBUG io.atomix.raft.impl.RaftContext - RaftServer{1} - Set term 1
20:01:13.785 [] TRACE io.atomix.raft.storage.system.MetaStore - Store vote 1
20:01:13.788 [] DEBUG io.atomix.raft.impl.RaftContext - RaftServer{1} - Voted for 1
20:01:13.789 [] INFO  io.atomix.raft.impl.RaftContext - RaftServer{1} - Transitioning to LEADER
20:01:13.800 [] INFO  io.atomix.raft.impl.RaftContext - RaftServer{1} - Found leader 1
20:01:13.801 [] TRACE io.atomix.raft.impl.RaftContext - RaftServer{1} - Set leader 1
20:01:13.874 [] TRACE io.atomix.raft.roles.LeaderRole - RaftServer{1}{role=LEADER} - Appended IndexedRaftLogEntryImpl{index=1, term=1, entry=InitialEntry{}, record=PersistedJournalRecord{metadata=RecordMetadata{checksum=3338092578, length=45}, record=RecordData{index=1, asqn=-1}}}
20:01:13.882 [] INFO  io.atomix.raft.impl.RaftContext - RaftServer{1} - Setting firstCommitIndex to 1. RaftServer is ready only after it has committed events upto this index
20:01:13.888 [] TRACE io.atomix.raft.roles.LeaderRole - RaftServer{1}{role=LEADER} - Starting append timer on fix rate of PT0.25S
20:01:13.895 [] INFO  io.atomix.raft.impl.DefaultRaftServer - RaftServer{1} - Server join completed. Waiting for the server to be READY
20:01:13.896 [] DEBUG io.atomix.raft.impl.DefaultRaftServer - RaftServer{1} - Server started successfully!
20:01:13.910 [] INFO  io.atomix.raft.impl.RaftContext - RaftServer{1} - Transitioning to FOLLOWER
20:01:13.913 [] TRACE io.atomix.raft.roles.LeaderRole - RaftServer{1}{role=LEADER} - Cancelling append timer
20:01:13.914 [] TRACE io.atomix.raft.impl.RaftContext - RaftServer{1} - Set leader null
20:01:13.914 [] INFO  io.atomix.raft.roles.FollowerRole - RaftServer{1}{role=FOLLOWER} - Single member cluster. Transitioning directly to candidate.
20:01:13.915 [] INFO  io.atomix.raft.impl.RaftContext - RaftServer{1} - Transitioning to CANDIDATE
20:01:13.916 [] INFO  io.atomix.raft.roles.CandidateRole - RaftServer{1}{role=CANDIDATE} - Single member cluster. Transitioning directly to leader.
20:01:13.916 [] TRACE io.atomix.raft.storage.system.MetaStore - Store term 2
20:01:13.920 [] TRACE io.atomix.raft.storage.system.MetaStore - Store vote null
20:01:13.923 [] DEBUG io.atomix.raft.impl.RaftContext - RaftServer{1} - Set term 2
20:01:13.923 [] TRACE io.atomix.raft.storage.system.MetaStore - Store vote 1
20:01:13.958 [] DEBUG io.atomix.raft.impl.RaftContext - RaftServer{1} - Voted for 1
20:01:13.959 [] INFO  io.atomix.raft.impl.RaftContext - RaftServer{1} - Transitioning to LEADER
20:01:13.960 [] INFO  io.atomix.raft.impl.RaftContext - RaftServer{1} - Found leader 1
20:01:13.960 [] TRACE io.atomix.raft.impl.RaftContext - RaftServer{1} - Set leader 1
20:01:13.961 [] TRACE io.atomix.raft.roles.LeaderRole - RaftServer{1}{role=LEADER} - Appended IndexedRaftLogEntryImpl{index=2, term=2, entry=InitialEntry{}, record=PersistedJournalRecord{metadata=RecordMetadata{checksum=2770027399, length=45}, record=RecordData{index=2, asqn=-1}}}
20:01:13.964 [] TRACE io.atomix.raft.roles.LeaderRole - RaftServer{1}{role=LEADER} - Starting append timer on fix rate of PT0.25S
20:01:13.967 [] INFO  io.atomix.raft.impl.RaftContext - RaftServer{1} - Transitioning to INACTIVE
20:01:13.967 [] TRACE io.atomix.raft.roles.LeaderRole - RaftServer{1}{role=LEADER} - Cancelling append timer
20:01:13.967 [] TRACE io.atomix.raft.impl.RaftContext - RaftServer{1} - Set leader null
20:01:13.971 [] DEBUG io.atomix.raft.impl.zeebe.LogCompactor - LogCompactor{1} - Closing the log compactor 1
20:01:13.974 [] DEBUG io.zeebe.journal.file.SegmentedJournal - Closing segment: JournalSegment{id=1, version=1, index=1}
</pre>
</details>

--- a/atomix/cluster/src/test/java/io/atomix/raft/SingleRaftEntryValidationTest.java
+++ b/atomix/cluster/src/test/java/io/atomix/raft/SingleRaftEntryValidationTest.java
@@ -19,9 +19,11 @@ import static org.assertj.core.api.Assertions.assertThat;
 import static org.assertj.core.api.Assertions.assertThatThrownBy;
 
 import io.atomix.raft.storage.log.entry.ApplicationEntry;
+import io.atomix.raft.storage.log.entry.InitialEntry;
 import io.atomix.raft.zeebe.EntryValidator;
 import io.atomix.raft.zeebe.ValidationResult;
-import java.util.List;
+import io.zeebe.test.util.TestUtil;
+import java.util.Collection;
 import java.util.function.BiFunction;
 import org.junit.Rule;
 import org.junit.Test;
@@ -50,17 +52,19 @@ public class SingleRaftEntryValidationTest {
     // when
     assertThatThrownBy(() -> raftRule.appendEntry()).hasMessageContaining("invalid");
     entryValidator.validation = (last, current) -> ValidationResult.success();
-    raftRule.awaitNewLeader();
-    final var commitIndex =
-        raftRule.appendEntry(); // append another entry to await the commit index
+    TestUtil.waitUntil(() -> getEntryTypeCount(InitialEntry.class) == 2);
+    raftRule.appendEntry();
 
     // then
-    raftRule.awaitCommit(commitIndex);
-    raftRule.awaitSameLogSizeOnAllNodes(commitIndex);
-    final var memberLog = raftRule.getMemberLogs();
+    assertThat(getEntryTypeCount(ApplicationEntry.class)).isOne();
+  }
 
-    final var logLength = memberLog.values().stream().map(List::size).findFirst().orElseThrow();
-    assertThat(logLength).withFailMessage(memberLog.toString()).isEqualTo(3);
+  private int getEntryTypeCount(final Class type) {
+    return (int)
+        raftRule.getMemberLogs().values().stream()
+            .flatMap(Collection::stream)
+            .filter(e -> e.entry().getClass().isAssignableFrom(type))
+            .count();
   }
 
   private static class TestEntryValidator implements EntryValidator {

